src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -722,8 +722,15 @@ private static int classIndex(Class<?> cl) {
     private static final int[] PREPEND_FILTER_SECOND_PAIR_ARGS = new int[] { 0, 1, 4, 5 };
 
     // Base MH for complex prepender combinators.
-    private static final MethodHandle PREPEND_BASE = MethodHandles.dropArguments(
-            MethodHandles.identity(long.class), 1, byte[].class);
+    private static @Stable MethodHandle PREPEND_BASE;
+    private static MethodHandle prependBase() {
+        MethodHandle base = PREPEND_BASE;
+        if (base == null) {
+            base = PREPEND_BASE = MethodHandles.dropArguments(
```

Comment:
I'm curious about this idiom - couldn't this result in the `PREPEND_BASE` field being assigned twice, with objects that are morally identical, but could have different identities?

Reply:
It's not a violation, it's relying on behavior not discussed by the `@Stable` doc, which refers to such behavior as "undefined".

It's a carefully correlated dance between the code, interpreter, and C2 relying on:
1. safe publication of equivalent immutable structures with no dependence on their identity when operating on them;
2. the field being set in the interpreter; and 
3. by the time C2 inlines `prependBase` the field is no longer be updated, thus C2 can replace the field access with its value.

Reply:
Thanks, I think I was missing (2)